google/cloud/aiplatform/docker_utils/build.py

```diff
@@ -0,0 +1,430 @@
+    requirements_path: Optional[str] = None,
+    setup_path: Optional[str] = None,
+    extra_packages: Optional[List[str]] = None,
+    container_workdir: Optional[str] = None,
+    container_home: Optional[str] = None,
+    extra_dirs: Optional[List[str]] = None,
+    exposed_ports: Optional[List[int]] = None,
+    pip_command: str = "pip",
+    python_command: str = "python",
+    no_cache: bool = True,
+    **kwargs,
+) -> Image:
+    """Builds a Docker image.
+
+    Generates a Dockerfile and passes it to `docker build` via stdin.
+    All output from the `docker build` process prints to stdout.
+
+    Args:
+        base_image (str):
+            Required. The ID or name of the base image to initialize the build stage.
+        host_workdir (str):
+            Required. The path indicating where all the required sources locates.
+        main_script (str):
+            Required. The string that identifies the executable script under the working
+            directory.
+        output_image_name (str):
+            Required. The name of the built image.
+        python_module (Optional[str]):
+            Optional. The executable main script in form of a python module, if applicable.
+        requirements (Optional[List[str]]):
+            Optional. The list of required dependencies to install from PyPI.
+        requirements_path (Optional[str]):
+            Optional. The path to a local file including required dependencies to install from PyPI.
+        setup_path (Optional[str]):
+            Optional. The path to a local setup.py used for installing packages.
+        extra_packages (Optional[List[str]]):
+            Optional. The list of user custom dependency packages to install.
+        container_workdir (Optional[str]):
+            Optional. The working directory in the container.
+        container_home (Optional[str]):
+            Optional. The $HOME directory in the container.
+        extra_dirs (Optional[List[str]]):
+            Optional. The directories other than the work_dir required.
+        exposed_ports (Optional[List[int]]):
+            Optional. The exposed ports that the container listens on at runtime.
+        pip_command (str):
+            Required. The pip command used for installing packages.
+        python_command (str):
+            Required. The python command used for running python scripts.
+        no_cache (bool):
+            Required. Do not use cache when building the image.
+        **kwargs:
+            Other arguments to pass to underlying method that generates the Dockerfile.
+
+    Returns:
+        A Image class that contains info of the built image.
+
+    Raises:
+        DockerError: An error occurred when executing `docker build`
+    """
+
+    tag_options = ["-t", output_image_name]
+    cache_args = ["--no-cache"] if no_cache else []
+
+    command = (
+        ["docker", "build"] + cache_args + tag_options + ["--rm", "-f-", host_workdir]
```

Comment:
Docker installation requires root access.
Why not use Google Cloud Build instead?
See https://github.com/Ark-kun/google_cloud_mlflow/blob/884ae51a889d6639af50d2958cce92131a2ba11e/google_cloud_mlflow/_mlflow_models_docker_utils_patch.py#L60

Reply:
We discussed that. We will also have docker run functionality so users still need to install docker locally. Cloud Build is another option but will not be P0.